### PR TITLE
Refactor "clean old client keys"

### PR DESCRIPTION
**Pull request by auumgn**

### CODE REVIEW

File: orcid-core/src/main/java/org/orcid/core/cron/CleanOldClientKeysCronJob.java [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+import org.orcid.persistence.jpa.entities.ClientSecretEntity;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.stereotype.Component;
+import org.springframework.transaction.annotation.Transactional;
+
+@Component
+public class CleanOldClientKeysCronJob {
+    
+    private static final Logger LOGGER = LoggerFactory.getLogger(ClientDetailsManagerImpl.class);
```
Comment (amontenegro) [inline orcid-core/src/main/java/org/orcid/core/cron/CleanOldClientKeysCronJob.java:24]: @auumgn please use CleanOldClientKeysCronJob.class instead in the logger

File: orcid-persistence/src/main/java/org/orcid/persistence/dao/impl/ClientSecretDaoImpl.java [hunk 2]
```diff
@@ -91,4 +91,18 @@ public boolean setAsPrimary(ClientSecretEntity clientSecret) {
         query.setParameter("clientSecret", clientSecret.getClientSecret());
         return query.executeUpdate() > 0;
     }
+    
+    @Override
+    @Transactional
+    public List<ClientSecretEntity> getNonPrimaryKeys() {
+        TypedQuery<ClientSecretEntity> query = entityManager.createQuery("From ClientSecretEntity WHERE is_primary = true and last_modified + '24 hours' > '2021-01-01'", ClientSecretEntity.class);
```
Comment (amontenegro) [inline orcid-persistence/src/main/java/org/orcid/persistence/dao/impl/ClientSecretDaoImpl.java:98]: @auumgn `is_primary = true` use `false`

File: orcid-core/src/main/java/org/orcid/core/cron/CleanOldClientKeysCronJob.java [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+        if (nonPrimaryKeys.size() > 0) {
+            int i = 0;
+            StringBuilder queryCondition = new StringBuilder();
+            List<String> clientIds = new ArrayList<String>();
+            for (ClientSecretEntity e : nonPrimaryKeys) {
+                i++;
+                ClientDetailsEntity clientDetails = e.getClientDetailsEntity();
+                String clientId = clientDetails.getId();
+                String key = e.getClientSecret();
+                String s = String.format("client_details_id = '%1$s' and client_secret = '%2$s'", clientId, key);
```
Comment (amontenegro) [inline orcid-core/src/main/java/org/orcid/core/cron/CleanOldClientKeysCronJob.java:52]: @auumgn please surround each entry in parentheses

File: orcid-core/src/main/java/org/orcid/core/cron/CleanOldClientKeysCronJob.java [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+                queryCondition.append(s);
+                clientIds.add(clientId);
+                if (i != nonPrimaryKeys.size()) queryCondition.append(" or ");
+            };
+            String queryConditionString = queryCondition.toString();
+            LOGGER.info("The following keys are going to be deleted:\n" + queryConditionString.replace(" =", ":").replace(" and", ",").replace(" or", "\n").replace("'",""));
```
Comment (amontenegro) [inline orcid-core/src/main/java/org/orcid/core/cron/CleanOldClientKeysCronJob.java:58]: Make this LOGGER.info a LOGGER.warn

**Review by amontenegro — CHANGES_REQUESTED** [review]

**Comment by auumgn** [comment]
[Optimize the cleanOldClientKeys Spring cron job](https://trello.com/c/uxD5ItKd/7470-optimize-the-cleanoldclientkeys-spring-cron-job)

File: orcid-core/src/test/java/org/orcid/core/cron/CleanOldClientKeysCronJobTest.java [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+        ClientDetailsEntity client = new ClientDetailsEntity("anything", "anything");
+        ClientSecretEntity entity = new ClientSecretEntity();
+        entity.setClientSecret("anything");
+        entity.setClientDetailsEntity(client);
+        list.add(entity);
+        Mockito.when(clientSecretDao.getNonPrimaryKeys()).thenReturn(list);
+        Mockito.when(clientSecretDao.removeWithCustomCondition(Mockito.anyString())).thenReturn(true);
+        CleanUpJob.cleanOldClientKeys();
+        Mockito.verify(clientSecretDao, Mockito.times(1)).getNonPrimaryKeys();
+        Mockito.verify(clientSecretDao, Mockito.times(1)).removeWithCustomCondition(condition.capture());
```
Comment (amontenegro) [inline orcid-core/src/test/java/org/orcid/core/cron/CleanOldClientKeysCronJobTest.java:61]: @auumgn lets get the content of that parameter and assert that it contains what you expects.

File: orcid-core/src/test/java/org/orcid/core/cron/CleanOldClientKeysCronJobTest.java [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+        Mockito.verify(clientDetailsDao, Mockito.times(1)).updateLastModifiedBulk(Mockito.anyList());
```
Comment (amontenegro) [inline orcid-core/src/test/java/org/orcid/core/cron/CleanOldClientKeysCronJobTest.java:62]: Evaluate what you capture.

File: orcid-core/src/test/java/org/orcid/core/cron/CleanOldClientKeysCronJobTest.java [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+        entityTwo.setClientSecret("anything again");
+        entityTwo.setClientDetailsEntity(client);
+        list.add(entityOne);
+        list.add(entityTwo);
+        Mockito.when(clientSecretDao.getNonPrimaryKeys()).thenReturn(list);
+        Mockito.when(clientSecretDao.removeWithCustomCondition(Mockito.anyString())).thenReturn(true);
+        CleanUpJob.cleanOldClientKeys();
+        Mockito.verify(clientSecretDao, Mockito.times(1)).getNonPrimaryKeys();
+        Mockito.verify(clientSecretDao, Mockito.times(1)).removeWithCustomCondition(condition.capture());
+        Mockito.verify(clientDetailsDao, Mockito.times(1)).updateLastModifiedBulk(Mockito.anyList());
```
Comment (amontenegro) [inline orcid-core/src/test/java/org/orcid/core/cron/CleanOldClientKeysCronJobTest.java:82]: Evaluate what you capture.

**Review by amontenegro — CHANGES_REQUESTED** [review]

File: orcid-persistence/src/main/java/org/orcid/persistence/dao/impl/ClientSecretDaoImpl.java [hunk 8]
```diff
@@ -91,4 +94,23 @@ public boolean setAsPrimary(ClientSecretEntity clientSecret) {
         query.setParameter("clientSecret", clientSecret.getClientSecret());
         return query.executeUpdate() > 0;
     }
+    
+    @Override
+    @Transactional
+    public List<ClientSecretEntity> getNonPrimaryKeys() {
+        DateTime dt = DateTime.now();
```
Comment (amontenegro) [inline orcid-persistence/src/main/java/org/orcid/persistence/dao/impl/ClientSecretDaoImpl.java:101]: Hey @auumgn, you are not updating this dt, so, it keeps being now(), please update it so it is now() - 24 hours before assign it to the "yesterday" variable.

Thanks

**Review by amontenegro — CHANGES_REQUESTED** [review]

**Review by amontenegro — APPROVED** [review]

**Pull request merged**